File: hw6/scala/task1.scala
import java.util.Date
import org.apache.spark.{SparkConf, SparkContext}
import org.json4s.jackson.JsonMethods.{compact, parse}

import java.io.{File, PrintWriter}
import scala.math.abs
import scala.util.Random

object task1 {

  def create_random_hash_funcs(n_bits:Int, n_hash_func: Int)= {
    val m = n_bits
    val r = Random
    var a_list:List[Long] = List()
    var b_list:List[Long] = List()
    while (a_list.length != n_hash_func) {
      val a = r.nextInt(500) + 1
      val b = r.nextInt(5000) + 100
      if (a % m != 0 && !a_list.contains(a)) {
        a_list = a_list :+ a.toLong
        b_list = b_list :+ b.toLong
      }
    }
    (a_list, b_list, m)
  }


  def hash_city(random_a: List[Long], random_b: List[Long], m: Int, city: String): List[Long] = {
    var hash_list:List[Long] = List()
    if (city.trim==""){
       return List(-1)
    }
    val x = abs(city.hashCode)
    for (i <- random_a.indices){
      val h_v = (random_a.apply(i) * x + random_b.apply(i)) % m

      hash_list = hash_list :+ h_v
    }
    hash_list
  }

  def is_exists(index_list: List[Long], bit_array: Array[Int]): String = {
    if (index_list.head == -1) {
      "0"
    } else {
      var exist = "1"
      for (index <- index_list) {
        if (bit_array(index.toInt) == 0) exist = "0"
      }
      exist
    }
  }

  def main(args: Array[String]): Unit = {
    val start_time = new Date().getTime
    val conf = new SparkConf().setMaster("local[3]").setAppName("task1")
    val sc = new SparkContext(conf)
    sc.setLogLevel("WARN")
    val first_json_path = args(0) //"E:\\Data\\inf553\\hw6\\resource\\asnlib\\publicdata\\business_first.json"
    val second_json_path = args(1) //"E:\\Data\\inf553\\hw6\\resource\\asnlib\\publicdata\\business_second.json"
    val output_file_path = args(2) //"task1_ans.csv"

    val rdd_first = sc.textFile(first_json_path).map(item => parse(item))
      .map(item => compact(item \ "city"))
    val rdd_second = sc.textFile(second_json_path).map(item => parse(item))
      .map(item => compact(item \ "city"))
//    rdd_first.take(5).foreach(println)
    val n_bits = 10000
    val n_hash_func = 3
    val hash_params = create_random_hash_funcs(n_bits, n_hash_func)
//    println(hash_params)
    val rdd_indices = rdd_first.filter(item=> item.trim !="").flatMap(x=>hash_city(hash_params._1, hash_params._2, hash_params._3, x)).distinct()
    val indices = rdd_indices.collect()
    var bit_array:Array[Int] = new Array[Int](n_bits)
    for (index <- indices) {
      bit_array(index.toInt) = 1
    }
    val result = rdd_second.map(x=>hash_city(hash_params._1, hash_params._2, hash_params._3, x))
      .map(x=>is_exists(x, bit_array)).collect()

    val pw = new PrintWriter(new File(output_file_path))
    pw.write(result.mkString(" "))
    pw.close()
    val end_time =new Date().getTime
    println("time: "+(end_time-start_time)/1000.0+"s")
  }
}

File: hw6/python/task1.py
import os
import sys
import csv
import time
import json
import random
import binascii
from pyspark import SparkContext

def create_random_hash_funcs(n_bits, n_hash_func):
    m = n_bits
    random_a = list()
    random_b = list()
    while len(random_a) != n_hash_func:
        a = random.randint(1, 5000)
        b = random.randint(1000, 5000000)
        if a % m != 0 and a not in random_a:
            random_a.append(a)
            random_b.append(b)
    return random_a, random_b, m

def hash_city(a_list, b_list, m, city):
    hashed_list = list()
    # print(city)
    if city.strip() == "":
        return [-1]
    x = int(binascii.hexlify(city.encode('utf8')),16)
    for i in range(len(a_list)):
        h_v = (a_list[i] * x + b_list[i]) % m
        hashed_list.append(h_v)
    return hashed_list


def is_exists(index_list, bit_list):
    if index_list[0] == -1:
        return "0"
    else:
        exist = "1"
        for index in index_list:
            if bit_list[index] == 0:
                exist = "0"
        return exist

if __name__ == '__main__':
    start_time = time.time()
    os.environ['JAVA_HOME'] = '/home/buyi/app/jdk'
    os.environ['PYSPARK_PYTHON'] = '/usr/bin/python3'
    os.environ['PYSPARK_DRIVER_PYTHON'] = '/usr/bin/python3'

    sc = SparkContext('local', 'task1')
    first_json_path, second_json_path, output_file_path = "/home/buyi/data/hw6/business_first.json", "/home/buyi/data/hw6/business_second.json", "task1_ans.csv"
    # first_json_path, second_json_path, output_file_path = sys.argv[1:]

    rdd_first = sc.textFile(first_json_path).map(json.loads).map(lambda x:x['city'])
    rdd_second = sc.textFile(second_json_path).map(json.loads).map(lambda x: x['city'])
    # m = rdd_first.count()  # num of elements inserted
    n_bits = 10000  # num of bits in array
    n_hash_func = 3
    random_a, random_b, m = create_random_hash_funcs(n_bits, n_hash_func)
    rdd_indices = rdd_first.filter(lambda x:x.strip()!="").flatMap(lambda x: hash_city(random_a, random_b, m, x)).distinct()
    indices = rdd_indices.collect()
    bit_list = [0]*n_bits
    for index in indices:
        bit_list[index] = 1
    result = rdd_second.map(lambda x:hash_city(random_a, random_b, m, x)).map(lambda x: is_exists(x, bit_list)).collect()

    # print(result.count())
    # print(result.filter(lambda x:x==0).count())
    print(len(result), rdd_second.count())

    with open(output_file_path, 'w') as f:
        # f.write(" ".join(result))
        writer = csv.writer(f, delimiter=' ')
        writer.writerow(result)
    print("time: {}s".format(time.time() - start_time))